File: Interview_Bit/Binary_Search/Matrix_Median.py
from bisect import bisect_right
class Solution:
    # @param A : list of list of integers
    # @return an integer
    def findMedian(self, A):
        r = len(A)
        c = len(A[0])
        mi = A[0][0]
        mx = 0
        for i in range(r):
            mi = min(mi,A[i][0])
            mx = max(mx,A[i][c-1])
        des = (r*c+1)//2
        while mi<mx:
            mid = (mi+mx)//2
            pl = [0]
            for i in range(r):
                j = bisect_right(A[i],mid)
                pl[0] = pl[0]+j
            if pl[0]<des:
                mi=  mid+1
            else:
                mx=mid
        return mi

File: Interview_Bit/Binary_Search/Matrix_Median.cpp
int Solution::findMedian(vector<vector<int> > &A) {
    if(A.empty())
        return 0;
    int n = A.size(), m = A[0].size();
    int mn = INT_MAX, mx = 0;
    for(int i=0; i<n; i++){
        mn = min(mn,A[i][0]);
        mx = max(mx,A[i][m-1]);
    }
    int mid = (n*m + 1)/2;
    while(mn<mx){
        int curr = mn + (mx-mn)/2;
        int count = 0;
        for(int i=0; i<n; i++){
            count += upper_bound(A[i].begin(),A[i].end(),curr) - A[i].begin();
        }
        if(count<mid)
            mn = curr + 1;
        else
            mx = curr;
    }
    return mn;
}
